### DISCUSSION TITLE
New sort option. Random

### DISCUSSION THREAD

**Discussion by noxamos** [gotson/komga#1558] [Ideas]
### Describe your suggested feature

I have a lot of comics and I want a way to allow new ones to surface instead of having to search though everything.

Other option is to have a random comic button which would take you to the details page of a random comic in the current library.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to the latest version.
- [X] I will fill out all of the requested information in this form.

**Comment by skymkmk** [comment]
The suggestion for this feature has been proposed for over a year, but there hasn't been any progress yet. It seems that either the people who need it or its priority are not very high 😢. 

Currently, I enjoy reading doujinshi by randomly picking one with my eyes closed to pass the time. It would be even better if there was a random sorting feature. I hope someone can submit a pull request or the author can consider it when he/she have the time. Thanks to author and all contributors for their efforts!

**Comment by skyrim663** [comment]
This feature is important for killing time on a daily basis, and it's also more fun to select comics when the library is very large. You don't have to see the same page every time you open it.

**Comment by profzelonka** [comment]
This feature is essential in comic discovery and choosing what to read instead of searching for a specific genre or title. Seems like a very easy thing to add, looking at the existing sort options. Applications like Plex have random - they also have the same options to refine filter down more and more, but not having a random way to scatter options really limits the filter results to having to view everything in alphabetical order, or in some sort order of file size, etc.

If we had this option we could be more open to reading various other comics we're not used to seeing. Not to mention how versatile the sorting can be between series, franchise, and even comics within series in the case of One-Shot type folders.

Please <3

**Comment by jamesleyD** [comment]
I've decided to work on implementing a shuffle mode for reading books from a series, as well as a random series selection feature. I'll keep you posted if I come up with anything interesting.

**Comment by jamesleyD** [comment]
### ***Hello***

On my version of the project, I've added a new random discovery feature, which includes :
1. A new icon to navigate to a random series (available only on the ```libraries/all/series``` page).
3. A new icon that lets you navigate to a random book once you're on a series page

There are **2 new APIs** as a result:
```GET api/v1/series/random``` : returns random series, default 5
```GET /api/v1/series/{seriesId}/books/random```: returns books in the series at random, default 5

They both can take a ```limit``` parameter, which is an integer that specifies the maximum number of elements to be retrieved at random. I've set a ceiling of 20 for this parameter for the time being. This limit is due to the use of the sql function ```random()``` which may be inefficient in large datasets. The best thing is to limit it or find a way not to use it, in my opinion.

***
Here's a demo:


https://github.com/gotson/komga/assets/77970325/bd8c55d8-dbf8-498c-bc28-dba6efabd516


***

It's certainly not a random sorting system, but maybe one of you would be interested. if you have any ideas to share don't hesitate. I don't think I'd do a PR, unless @gotson comes by and confirms that what I've done is viable.

<sub>English is not my first language, so I apologize if anything is unclear.</sub>

**Comment by skymkmk** [comment]
@jamesleyD This seems to be just doing a shuffle within a single page, rather than a global random sorting.

You are right; Performance should especially be a concern when the amount of series is huge. Perhaps we could retrieve all the metadata from the database into the JVM memory for random sorting? I haven't read the project code and haven't seen how other sorts are implemented, but there's no reason why global sorting based on any arbitrary field should be possible, yet random sorting would impact performance.

I'm also very curious about Gotson's attitude towards this enhancement. Anyway, thank you very much for initially implementing this feature, unlike me who has been lazy and hasn't even read any code XD.

**Comment by jamesleyD** [comment]
The project uses a sqlite database. In some databse like mysql, the random function can take a seed, which allows you to obtain constant random results and thus to paginate. The sqlite RANDOM() function doesn't take a seed. This makes random sorting complicated to implement with pagination. We would need to find a way of preserving the state of the pagination during random sorting. I'm sure there must be a way to get the same random sequence every time. Fetching all items at once is not a very optimized solution. After looking into the alternatives, I came to the conclusion that it's too much work xD

**Comment by DieselTech** [comment]
> retrieve all the metadata from the database into the JVM memory for random sorting?

Please no. My database is so large it's the reason that data isn't shown anymore when searching. I really don't think it could handle that kind of strain.

**Comment by gotson** [comment]
🎉 This issue has been resolved in `1.16.0` ([Release Notes](https://github.com/gotson/komga/releases/tag/1.16.0))
